jupyterhub/roles.py

```diff
@@ -75,35 +76,30 @@ def expand_self_scope(name):
 
 def _get_scope_hierarchy():
     """
-    Returns a dictionary of scopes:
-    scopes.keys() = scopes of highest level and scopes that have their own subscopes
-    scopes.values() = a list of first level subscopes or None
+    Returns:
+      scope hierarchy (dict): dictionary of available scopes and their hierarchy where
+        scopes.keys() = top level scopes and scopes that have their own subscopes
+        scopes.values() = list of immediate subscopes or None
     """
+    subscope_lists = [
+        value['subscopes']
+        for value in scopes.scope_definitions.values()
+        if 'subscopes' in value
+    ]
 
-    scopes = {
-        'self': None,
-        'all': None,
-        'admin:users': ['admin:users:auth_state', 'users'],
-        'users': ['read:users', 'users:activity'],
-        'read:users': [
-            'read:users:name',
-            'read:users:groups',
-            'read:users:activity',
-        ],
-        'users:activity': ['read:users:activity'],
-        'users:tokens': ['read:users:tokens'],
-        'admin:users:servers': ['admin:users:server_state', 'users:servers'],
-        'users:servers': ['read:users:servers'],
-        'read:users:servers': ['read:users:name'],
-        'admin:groups': ['groups'],
-        'groups': ['read:groups'],
-        'read:services': None,
-        'read:hub': None,
-        'proxy': None,
-        'shutdown': None,
-    }
+    scope_hierarchy = {}
+    for scope in scopes.scope_definitions.keys():
 
-    return scopes
+        has_subscopes = scopes.scope_definitions[scope].get('subscopes')
+        is_subscope = any(scope in subscope_list for subscope_list in subscope_lists)
+
+        if has_subscopes:
+            scope_hierarchy[scope] = scopes.scope_definitions[scope]['subscopes']
+        else:
+            if not is_subscope:
```

Comment:
What's the case for `not has_scopes and is_subscope`? I'm not sure why we need to know if a scope *is* a subscope or not.

Reply:
The scope hierarchy dict does not have the lowest level scopes as keys (which do not have subscopes themselves but are subscopes of other scopes), e.g., `read:users:name`, so it won't include those as keys. On the other hand, it still needs to include top level scopes that do not have any subscopes, e.g., `shutdown`. 

This is the format that is used to get subscopes onwards, which thinking of now is probably confusing. We could change it to include all scopes and adjust the `_expand_scope()` function.

Reply:
`_expand_scope` already needs to handle proxy/read:hub/etc. with `: None` right? So would it work to have None there for all leaf scopes?

Reply:
We do not actually need this at all! `scopes.scope_definitions` can be used directly (vis last commit).